### PR TITLE
Completion should detect the collection initializer pattern on types …

### LINKED ISSUE #4754: Object initializer completion should include readonly properties have WinRT projected Add methods
Create a C# UWP app.

In App.xaml.cs type:

``` C#
            var stackPanel = new StackPanel()
            {
                Children =
                {
                    new TextBlock { Text = "Hi" }
                }
            };
```

The `Children` property won't show in completion because the IDE doesn't know about it's add method, but the compiler accepts it.

### PR DESCRIPTION
**Pull request by rchande**
…that inherit it.

Fixes #4754.

cc: @Pilchie @dpoeschl @balajikris @jasonmalinowski @jmarolf @brettfo @CyrusNajmabadi @DustinCampbell  @davkean

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs [hunk 1]
```diff
@@ -612,6 +612,49 @@ End Property
                 hideAdvancedMembers: false);
         }
 
+        [WorkItem(4754, "https://github.com/dotnet/roslyn/issues/4754")]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: Can you add tests where the member is protected, and you either are/aren't in a subclass that has access to them?

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 2]
```diff
@@ -607,11 +607,11 @@ public static bool IsSpecialType(this ITypeSymbol symbol)
             return false;
         }
 
-        public static bool CanSupportCollectionInitializer(this ITypeSymbol typeSymbol)
+        public static bool CanSupportCollectionInitializer(this ITypeSymbol typeSymbol, ISymbol within)
         {
             return
                 typeSymbol.AllInterfaces.Any(i => i.SpecialType == SpecialType.System_Collections_IEnumerable) &&
-                typeSymbol.GetMembers(WellKnownMemberNames.CollectionInitializerAddMethodName)
+                typeSymbol.GetAccessibleMembersInThisAndBaseTypes<IMethodSymbol>(within ?? typeSymbol).Where(s => s.Name == WellKnownMemberNames.CollectionInitializerAddMethodName)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: why the ??   when are you not "within" at least a type or assembly?

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 3]
```diff
@@ -607,11 +607,11 @@ public static bool IsSpecialType(this ITypeSymbol symbol)
...
                     .OfType<IMethodSymbol>()
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: OfType<IMethodSymbol> seems redundant now.

**Comment by Pilchie** [comment]
Does this actually fix the WinRT case?  I thought the issue there was the that We don't actually see the "Add" method on the concrete type, but the compiler still knows how to call it? @agocke

**Comment by agocke** [comment]
I'm not sure I understand the bug -- why would `Children` not be visible? It's looks like a regular method inherited from Panel to me.

I can see how the collection initializer could be misread, but the bug doesn't explain how they're connected.

**Comment by rchande** [comment]
@Pilchie  I don't think this has anything to do with WinRT--the IDE didn't know how to check for the pattern if it was inherited. That was the problem with StackPanel and the scenario you describe (in UWP projects) works with this fix.

**Comment by Pilchie** [comment]
Interesting.  Maybe I saw a different bug while looking at this. I'm pretty sure Metadata as source didn't show me an add method in any ancestor, possibly because it was a projected special WinRT type.

**Comment by agocke** [comment]
@Pilchie That sounds about right. I suspect Metadata-as-source attempts to show the symbol view, but projections are only handled in Lookup. I would probably consider that a separate bug.

**Comment by Pilchie** [comment]
Agreed that it's a separate bug, it just led me to mis-diagnose this one :)

**Comment by Pilchie** [comment]
Wait - @rchande - this is showing as "Closed", not "Merged".  What happened?

**Pull request closed without merging**